src/Haskell/Ide/Engine/LSP/Reactor.hs

```diff
@@ -81,31 +79,38 @@ reactorSend' f = do
 
 -- ---------------------------------------------------------------------
 
+-- | Sends a single request to the scheduler so it can be be processed
+-- asynchronously.
 makeRequest :: (MonadIO m, MonadReader REnv m) => PluginRequest R -> m ()
-makeRequest req@(GReq _ _ Nothing (Just lid) _ _) = writePluginReq req lid
-makeRequest req@(IReq _ lid _ _) = writePluginReq req lid
-makeRequest req = liftIO . atomically . flip writeTChan req =<< asks reqChanIn
-
-writePluginReq :: (MonadIO m, MonadReader REnv m) => PluginRequest R -> J.LspId -> m ()
-writePluginReq req lid = do
-  wipTVar <- asks (wipReqsTVar . dispatcherEnv)
-  cin     <- asks reqChanIn
-  liftIO $ atomically $ do
-    modifyTVar wipTVar (S.insert lid)
-    writeTChan cin req
+makeRequest req = do
+  sc <- asks scheduler
+  liftIO $ Scheduler.sendRequest sc Nothing req
+
+-- | Updates the version of a document and then sends the request to be processed
+-- asynchronously.
+updateDocumentRequest
+  :: (MonadIO m, MonadReader REnv m) => Uri -> Int -> PluginRequest R -> m ()
+updateDocumentRequest uri ver req = do
+  sc <- asks scheduler
+  liftIO $ Scheduler.sendRequest sc (Just (uri, ver)) req
+
+-- | Marks a s requests as cencelled by its LspId
+cancelRequest :: (MonadIO m, MonadReader REnv m) => J.LspId -> m ()
```

Comment:
I would expect these functions (all the ones that do `asks scheduler` to move to the `Scheduler` module.
Would that make sense?

Reply:
Yes, but I suspect the only constraint we really need is that the `asks` call succeeds.

To me it is about putting all the stuff related to scheduling in one place, so we can look at it holistically.

Reply:
How would that look?

Reply:
And looking more closely, that is exactly what you proposed. Sorry.

Reply:
@alanz I moved the implementation of those functions to the `Scheduler` module. Decided to keep the same functions in the `Reactor` module in order to provide some more specific type signatures. Using the generic function from `Scheduler` in `LspStdio` required adding many annoying type hints. In my opinion this version looks cleaner.